- ~Remove other dataframes in beginning I dont use~
- Add introduction
    - Set narrative
        - Point to best alteranatives (1968 pdf, ???)
    - Introduce sources
    - Set expecations for data (124,000 figure from star tribune, figures from blog article)
    - Definitions (what is a lake?)
- Intersperse code with descriptions of what it's doing and add narrative
- Cite sources and add bibliography

- refactor into pipeline? this removes ability to explain the process. Maybe have seperate notebooks. This one for the data cleaning process.
- do something about mention of innaccuracy. remove it?

https://www.chrisfinke.com/2013/12/30/does-minnesota-really-have-more-shoreline-than-california/
https://resources.gisdata.mn.gov/pub/gdrs/data/pub/us_mn_state_dnr/water_dnr_hydrography/metadata/dnr_hydrography_lakes_and_open_water.html
https://gisdata.mn.gov/dataset/water-dnr-hydrography
https://www.dnr.state.mn.us/faq/mnfacts/water.html

In [1]:
import pandas as pd
import numpy as np

# show all columns
pd.set_option('display.max_columns', None)

# MN DNR
https://resources.gisdata.mn.gov/pub/gdrs/data/pub/us_mn_state_dnr/water_dnr_hydrography/metadata/dnr_hydrography_lakes_and_open_water.html

Downloaded ESRI File database from here (https://gisdata.mn.gov/dataset/water-dnr-hydrography) and loaded into Qgis then exported the feature table into an excel file and converted to csv. It might be possible to use python instead of Qgis gui to get feature table.

In [2]:
df = pd.read_csv('mn-dnr-hydro-features.csv')
df.head()

objectid     fw_id    dowlknum pw_basin_name                pw_parent_name  \
0         1   17992.0  23004406.0       Sparrow  Spring Valley Hatchery Ponds   
1         2  110905.0         NaN    Leaf River                    Leaf River   
2         3       NaN         NaN           NaN                           NaN   
3         4       NaN         NaN           NaN                           NaN   
4         5       NaN         NaN           NaN                           NaN   

  pw_sub_name sub_flag            wb_class  lake_class        acres  \
0     Sparrow        Y  Fish Hatchery Pond         NaN     0.343256   
1         NaN        R    Riverine polygon         0.0  1005.953305   
2         NaN      NaN        Lake or Pond         NaN     0.120631   
3         NaN      NaN        Lake or Pond         NaN     0.129182   
4         NaN      NaN        Lake or Pond         NaN     0.111783   

    shore_mi  center_utm_x  center_utm_y  cty_name  dnr_region fsh_office  \
0   0.087870   550665.9180   4837989.830  Fillmore         3.0        NaN   
1  66.240951           NaN           NaN       NaN         0.0        NaN   
2   0.051406   255405.9264   5311739.598       NaN         NaN        NaN   
3   0.051418   242472.6124   5315056.041       NaN         NaN        NaN   
4   0.051450   250616.0976   5387773.260       NaN         NaN        NaN   

  outside_mn delineated_by delineation_date      delineation_src  \
0          N           NaN              NaN        ArcGIS Online   
1          N           NaN              NaN  Buffered centerline   
2          N           NaN              NaN        DNR 24k POLYS   
3          N           NaN              NaN        DNR 24k POLYS   
4          N           NaN              NaN        DNR 24k POLYS   

                       delineation_notes approved_by approval_date  \
0  Re-delineation by DNR Fisheries staff    lybergqu           NaN   
1                                    NaN         NaN           NaN   
2                                    NaN         NaN           NaN   
3                                    NaN         NaN           NaN   
4                                    NaN         NaN           NaN   

         approval_notes has_flag flag_type         publish_date  \
0  name updated per LSM        Y      AQ18  25/03/2021 00:00:00   
1                   NaN        Y       NaN  25/03/2021 00:00:00   
2                   NaN      NaN       NaN  25/03/2021 00:00:00   
3                   NaN      NaN       NaN  25/03/2021 00:00:00   
4                   NaN      NaN       NaN  25/03/2021 00:00:00   

   lksdb_basin_id has_wld_flag wld_flag_type  unique_id created_user  \
0             NaN          NaN           NaN   145461.0     DNRHYDRO   
1             NaN          NaN           NaN        NaN     DNRHYDRO   
2             NaN          NaN           NaN   114730.0          NaN   
3             NaN          NaN           NaN   114672.0          NaN   
4             NaN          NaN           NaN   114234.0          NaN   

          created_date last_edited_user     last_edited_date ow_use pwi_class  \
0  01/09/2016 20:47:46         DNRHYDRO  18/05/2020 22:47:40      Y       NaN   
1  21/07/2016 23:44:25         DNRHYDRO  15/05/2020 16:33:14      Y       NaN   
2                  NaN         DNRHYDRO  03/12/2019 01:12:44      Y       NaN   
3                  NaN         DNRHYDRO  03/12/2019 00:56:08      Y       NaN   
4                  NaN         DNRHYDRO  03/12/2019 01:19:58      Y       NaN   

  map_display   map_label   shape_Length    shape_Area  \
0           Z     Sparrow     141.412697  1.389109e+03   
1           N  Leaf River  106604.690100  4.070949e+06   
2           N         NaN      82.730534  4.881747e+02   
3           N         NaN      82.749176  5.227801e+02   
4           N         NaN      82.801708  4.523686e+02   

                              unique_guid     INSIDE_X     INSIDE_Y  \
0  {3A96C1D6-6B7E-4E33-81CC-CF08BF1EC5F6}  550665.918

I will select the features that we will use. Many of the features are related to drawing polygons on a map.

In [3]:
features = [
    'fw_id', 'pw_basin_name', 'pw_parent_name', 'pw_sub_name',
    'sub_flag', 'wb_class', 'acres', 'shore_mi', 'cty_name'
]
df = df[features]
df.head()

fw_id pw_basin_name                pw_parent_name pw_sub_name sub_flag  \
0   17992.0       Sparrow  Spring Valley Hatchery Ponds     Sparrow        Y   
1  110905.0    Leaf River                    Leaf River         NaN        R   
2       NaN           NaN                           NaN         NaN      NaN   
3       NaN           NaN                           NaN         NaN      NaN   
4       NaN           NaN                           NaN         NaN      NaN   

             wb_class        acres   shore_mi  cty_name  
0  Fish Hatchery Pond     0.343256   0.087870  Fillmore  
1    Riverine polygon  1005.953305  66.240951       NaN  
2        Lake or Pond     0.120631   0.051406       NaN  
3        Lake or Pond     0.129182   0.051418       NaN  
4        Lake or Pond     0.111783   0.051450       NaN

Some of these features are not lakes. Let's get the count of each unique feature in the `wb_class` column. 

In [4]:
df["wb_class"].value_counts()

Lake or Pond            117420
Wetland                   4011
Intermittent Water        3520
Island or Land            1148
Riverine island           1136
Riverine polygon           843
Drained Wetland            820
Artificial Basin           481
Innundation Area           395
Drained Lakebed            287
Sewage/Filtration Pd       255
P. Drained Wetland         170
Natural Ore Mine           103
P. Drained Lakebed          76
Fish Hatchery Pond          75
Mine Pit Lake               62
Reservoir                   38
Mine Pit Lake (NF)          22
Tailings Pond               18
Mine or Gravel Pit          11
Industrial Waste Pd          4
?                            2
Tac/Natural Ore Mine         1
Name: wb_class, dtype: int64

So according to this dataset, there are 117,420 lakes or ponds in Minnesota. This is slightly less than the 124,400 reported by Stephen Aichele, a USGS geographer, in this Star Tribune Article. https://www.startribune.com/wisconsin-claims-it-has-more-lakes-than-minnesota-spoiler-alert-it-doesn-t/511709092/

This is only about 6% less, and Aichele referenced a different dataset. So overall, this number makes sense.

We still need to narrow down the data a bit more. The `sub_flag` column indicates that the record is a child of another record. This can be exemplified by search for "Minnetonka" in the `pw_basin_name` column.

In [5]:
# select rows that are "Lake or Pond" only then drop wb_class
df = df[df["wb_class"] == "Lake or Pond"]
df = df.drop(["wb_class"], axis=1)

minnetonka = df[df["pw_basin_name"].str.contains("Minnetonka", na=False)].head()
minnetonka.head()

fw_id            pw_basin_name pw_parent_name  \
18929  88888.0     Minnetonka-North Arm     Minnetonka   
23721  88888.0    Minnetonka-Upper Lake     Minnetonka   
44820  88888.0   Minnetonka-Carsons Bay     Minnetonka   
44834  88888.0   Minnetonka-Maxwell Bay     Minnetonka   
44844  88888.0  Minnetonka-Halsteds Bay     Minnetonka   

                   pw_sub_name sub_flag        acres   shore_mi  cty_name  
18929     Minnetonka-North Arm        Y   324.402774   5.657976  Hennepin  
23721    Minnetonka-Upper Lake        Y  4257.355511  34.196433  Hennepin  
44820   Minnetonka-Carsons Bay        Y   114.612685   3.283466  Hennepin  
44834   Minnetonka-Maxwell Bay        Y   304.687898   4.081228  Hennepin  
44844  Minnetonka-Halsteds Bay        Y   571.559165   8.118689  Hennepin

We can safely discard these sub records (bays, arms, etc.).

In [6]:
# Select all where sub_flag is not Y - this includes many NaN
df = df[df["sub_flag"] != "Y"]
df.head()

fw_id pw_basin_name pw_parent_name pw_sub_name sub_flag     acres  \
2    NaN           NaN            NaN         NaN      NaN  0.120631   
3    NaN           NaN            NaN         NaN      NaN  0.129182   
4    NaN           NaN            NaN         NaN      NaN  0.111783   
5    NaN           NaN            NaN         NaN      NaN  0.905717   
6    NaN           NaN            NaN         NaN      NaN  0.214522   

   shore_mi cty_name  
2  0.051406      NaN  
3  0.051418      NaN  
4  0.051450      NaN  
5  0.173349      NaN  
6  0.074835      NaN

We can see that there are a lot of NaN values.

In [7]:
# number of NaN values in sub_flag
df['sub_flag'].isna().sum()

93768

In [8]:
# number of N values in sub_flag
len(df[df['sub_flag'] == 'N'])

23044

This figure is similar to the number of lakes Finke got from the DNR. I want to explore these `NaN` values a bit more. I expect most of these are small and unnamed.

### sub_flag == NaN

#### Size

In [9]:
sub_flag_na = df[df['sub_flag'].isna()]
sub_flag_na['acres'].describe()

count    93768.000000
mean         0.951991
std          3.121067
min          0.102810
25%          0.203757
50%          0.342877
75%          0.865662
max        525.178383
Name: acres, dtype: float64

The mean acreage is less than 1. This means the average lake in this subset is about 10% of the size necessary to be considered a lake.

#### Names

In [10]:
sub_flag_na['pw_basin_name'].value_counts()

Unnamed              3
West Wing            1
Ewoldt Pond          1
Pepper               1
Owl Pond             1
Phyllis Pond         1
East Wing            1
Wild Prairie Pond    1
McGee Pond           1
Name: pw_basin_name, dtype: int64

This shows that out of the 90,000+ lakes, only a few have names (including 'unnamed').

### sub_flag ==  N
#### Size

In [11]:
sub_flag_notna = df[df['sub_flag'].notna()]['acres']
sub_flag_notna.describe()

count    2.304400e+04
mean     1.041920e+03
std      1.241164e+05
min      4.369010e-03
25%      6.085234e+00
50%      1.364216e+01
75%      4.396664e+01
max      1.876000e+07
Name: acres, dtype: float64

The mean for lakes that difference is pretty significant. Now that we have removed the sub basins and characterized them we can remove the related features.

In [12]:
df = df.drop(["pw_parent_name","pw_sub_name","sub_flag"], axis=1)

In [13]:
# This returns the largest NaN lakes. The largest of these is actually not a lake!
# If you take the coordinates and view it in QGIS then find this same location on Google maps satellite view
# then you can see it is just a field! Interestingly, google maps regular view shows it as a lake as well.
# df[df['sub_flag'].isna()].sort_values('acres').tail()

When sorting by acreage we can see a few lakes that are accounted for twice. These large lakes are on the border of Minnesota. One row shows the portion in Minnesota, and the other shows the portion in the other regions. 

In [14]:
df.sort_values(by=["acres"])

fw_id               pw_basin_name         acres     shore_mi  \
67802   17612.0                   Nemo Pond  4.369010e-03     0.009525   
1586    12148.0                     Unnamed  1.436461e-02     0.039564   
18794   17751.0                MaryMac Pond  1.495456e-02     0.017887   
69368   17270.0           Logging Camp Pond  1.559474e-02     0.018285   
90818   17747.0                Kenneth Pond  1.663244e-02     0.018583   
...         ...                         ...           ...          ...   
11621       NaN                         Red  2.838146e+05   127.118654   
130860   3738.0      Lake of the Woods (MN)  3.055776e+05   234.471821   
113543  99923.0  Lake of the Woods (Canada)  4.200037e+05  3047.677137   
39287     282.0                    Superior  1.625846e+06   407.487667   
125089  99999.0           Superior (WI, MI)  1.876000e+07  3387.876010   

                 cty_name  
67802              Isanti  
1586              Douglas  
18794             Stearns  
69368           St. Louis  
90818             Stearns  
...                   ...  
11621            Beltrami  
130860  Lake of the Woods  
113543          NOT IN MN  
39287                Cook  
125089          NOT IN MN  

[116812 rows x 5 columns]

We could filter by `cty_name`, but the DNR documentation on attributes shows that the `fw_id` for non-Minnesotan portion of border waters will be `99999` and 'alternate' (duplicate) representations of basins are `88888`. So we can select everything with a `fw_id` less than `88888` (or `NaN`) to exclude both of these. https://resources.gisdata.mn.gov/pub/gdrs/data/pub/us_mn_state_dnr/water_dnr_hydrography/metadata/dnr_hydrography_lakes_and_open_water.html

In [15]:
# select records by fw_id and drop feature
df = df[(df['fw_id'] < 88888) | (df['fw_id'].isna())]
df = df.drop(['fw_id'], axis=1)
len(df)

116710

At this point we have a complete list of lakes and ponds. We have removed all other basin types, excluded some duplicate and out-of-bounds records, and removed sub basins. We can cleanup the headings, reindex and export a full version and lake version (records >= 10 acres).

In [16]:
new_columns = {
    'pw_basin_name': 'name', 
    'shore_mi': 'shore_miles', 
    'cty_name': 'county'
}
df = df.rename(columns= new_columns)

In [17]:
df = df.reset_index(drop=True)
df.to_csv('lakes-and-ponds.csv')

In [18]:
lakes = df[df["acres"] >= 10]
lakes = lakes.reset_index(drop=True)
lakes.to_csv('lakes.csv')

This number is a bit larger than I would expect given that the Minnesota DNR officially recognizes 11,842 lakes.

In [19]:
lakes.sort_values(by='acres')

name         acres  shore_miles             county
13510                 Unnamed  1.000034e+01     0.547027            Stearns
11558                     NaN  1.000279e+01     0.846167                NaN
5423                  Unnamed  1.000304e+01     0.526680              Swift
8524                  Unnamed  1.000342e+01     0.466297           Le Sueur
6517                     Echo  1.000694e+01     0.488595               Cass
...                       ...           ...          ...                ...
6660                    Leech  1.030396e+05   186.231660               Cass
3681               Mille Lacs  1.282505e+05   105.472554         Mille Lacs
1226                      Red  2.838146e+05   127.118654           Beltrami
14053  Lake of the Woods (MN)  3.055776e+05   234.471821  Lake of the Woods
4022                 Superior  1.625846e+06   407.487667               Cook

[14062 rows x 4 columns]

In [20]:
notNa = lakes["name"].notna()
notUnnamed = lakes["name"].str.contains("Unnamed")
lakes[(notNa) | (notUnnamed)]

name       acres  shore_miles    county
0           Bugo   30.377878     1.016194      Lake
1       Pea Soup   12.101236     0.623966      Lake
2        Sundown   19.021747     0.721198      Lake
3        Unnamed   18.746844     0.633619    Aitkin
4      Studhorse   19.658274     0.722163    Aitkin
...          ...         ...          ...       ...
14057      Grave  372.807961     6.189930      Cass
14058     Jensen   52.293865     2.075376    Dakota
14059      Welch  196.064298     2.230892      Cass
14060     German  888.773305     7.148789  Le Sueur
14061       Long  131.010846     4.037602    Itasca

[13353 rows x 4 columns]